src/persist/src/indexed/runtime.rs

```diff
@@ -211,16 +222,97 @@ impl RuntimeClient {
     /// This snapshot is guaranteed to include any previous writes.
     ///
     /// The id must have previously been registered.
-    pub fn snapshot(&self, id: Id, res: CmdResponse<IndexedSnapshot>) {
+    fn snapshot(&self, id: Id, res: CmdResponse<IndexedSnapshot>) {
         self.core.send(Cmd::Snapshot(id, res))
     }
 
-    /// Synchronously stops the runtime, causing all future commands to error.
+    /// Synchronously closes the runtime, releasing exclusive-writer locks and
+    /// causing all future commands to error.
     ///
     /// This method is idempotent.
     pub fn stop(&mut self) -> Result<(), Error> {
         self.core.stop()
     }
+
+    /// Remove the persisted stream.
+    ///
+    /// TODO: Should this live on Meta?
+    pub fn destroy(&mut self, _id: &str) -> Result<(), Error> {
+        // TODO: When we implement this, we'll almost certainly want to put both
+        // the external string stream name and internal u64 stream id into a
+        // graveyard, so they're not accidentally reused.
+        unimplemented!()
+    }
+}
+
+/// A handle that allows writes of ((Key, Value), Time, Diff) updates into an
+/// [crate::indexed::Indexed] via a [RuntimeClient].
+pub struct StreamWriteHandle {
+    id: Id,
+    runtime: RuntimeClient,
+}
+
+impl StreamWriteHandle {
+    /// Returns a new [StreamWriteHandle] for the given stream.
+    pub fn new(id: Id, runtime: RuntimeClient) -> Self {
+        StreamWriteHandle { id, runtime }
+    }
+
+    /// Synchronously writes (Key, Value, Time, Diff) updates.
+    pub fn write_sync(&mut self, updates: &[((String, String), u64, isize)]) -> Result<(), Error> {
+        // TODO: Make Write::write_sync signature non-blocking.
+        let (rx, tx) = mpsc::channel();
+        self.runtime.write(self.id, updates, rx.into());
+        tx.recv().map_err(|_| Error::RuntimeShutdown)?
+    }
+
+    /// Closes the stream at the given timestamp, migrating data strictly less
+    /// than it into the trace.
+    pub fn seal(&mut self, upper: u64) -> Result<(), Error> {
+        // TODO: Make Write::write_sync signature non-blocking.
+        let (rx, tx) = mpsc::channel();
+        self.runtime.seal(self.id, upper, rx.into());
+        tx.recv().map_err(|_| Error::RuntimeShutdown)?
+    }
+}
+
+/// A handle for a persisted stream of ((Key, Value), Time, Diff) updates backed
+/// by an [crate::indexed::Indexed] via a [RuntimeClient].
+pub struct StreamMetaHandle {
```

Comment:
does it make sense to call this a `StreamReadHandle`? `allow_compaction` is broadly the same as "I don't care about times less than <X>" which still seems to fit, and `snapshot` is obviously a read op

Reply:
Yeah that's fine by me